### ISSUE TITLE
Include GPU daemonset in GKE configs?

### ISSUE THREAD

**Issue by jlewi** [kubeflow/kubeflow#288]
On GKE users need to deploy a daemonset to configure GPU nodes

```
https://cloud.google.com/kubernetes-engine/docs/concepts/gpus#installing_drivers
```

I think the main advantage is eliminating another step users have to do get a GKE cluster configured.

I'm guessing we can probably already start the daemonset as it only runs on GPU nodes so starting it on a CPU cluster is a null op.

I'm wondering if having multiple copies of the daemonet is actually problematic (e.g. 1 started manually and 1 via Kubeflow). My guess is one of them would succeed on a given node and mark the node as processed.

One problem could be is if the daemonsets are installing different versions of the drivers users might want up with a random mix of driver versions on nodes.

/cc @vishh

**Comment by bhack** [comment]
Seems that we are quite stalled for other providers or very frequent Kops deployments on different providers https://github.com/kubernetes/kubernetes/issues/54011

**Comment by jlewi** [comment]
Thanks @bhack.

**Comment by swiftdiaries** [comment]
Just a thought, I have a script that installs drivers manually. [Install drivers](https://github.com/swiftdiaries/dl-kops/blob/master/scripts/setupgpu.sh).
Additionally, it also installs CUDA and CuDNN. 
We could version it using environment variables and containerize it to deploy it as daemon set (using ksonnet) with a default set of parameters. This could be editable using the 'ks param set', so that users can choose to install different versions if they want to. 
This can be done on any Ubuntu machine and not just GKE.
The script is a bit of a mess, but any thoughts on the idea?

**Comment by bhack** [comment]
See also https://github.com/GoogleCloudPlatform/container-engine-accelerators/pull/51 and https://github.com/kubernetes/kops/pull/4971

**Comment by lluunn** [comment]
Can we just always deploy the daemonset as it has tolerations on gpu:
https://github.com/GoogleCloudPlatform/container-engine-accelerators/blob/master/daemonset.yaml#L45?

**Comment by bhack** [comment]
https://github.com/GoogleCloudPlatform/container-engine-accelerators/pull/51#issuecomment-399187623

**Comment by lluunn** [comment]
Thanks @bhack , looks really cool.

IIUC, using this, we can just run a (driver) container and avoid installing the driver?
Is this going to be the suggested pattern?

Paste the link here:
https://docs.google.com/presentation/d/1NY4X2K6BMaByfnF9rMEcNq6hS3NtmOKGTfihZ44zfrw/edit#slide=id.g3730f1de4f_0_16
https://github.com/NVIDIA/nvidia-container-runtime/

**Comment by jlewi** [comment]
@lluunn Can we just follow whatever GKE recommends and then update it when GKE updates their solution?

**Comment by bhack** [comment]
Why GKE only?

**Comment by jlewi** [comment]
This issue is GKE specific. I don't think Kubeflow has the bandwidth to try to figure out the proper way to install and configure GPUs on different K8s distributions. This should be solved upstream in Kubernetes.

GPUs are supported in GKE. The sole purpose of this issue is to remove the extra step of making users deploy the daemonset to install GPU drivers on GKE clusters.
https://cloud.google.com/kubernetes-engine/docs/concepts/gpus#installing_drivers

**Comment by bhack** [comment]
It is not so fair IMHO.
We could just point users to the offcial [Kubernetes documentation](https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/) without including specific partial solution in the repository.

Don't seems to you a little bit more vendor neutral?

**Comment by jlewi** [comment]
We explicitly want to provide mechanisms for vendors to optimize the experience for their particular platform. So in this particular case we want to provide a good deployment experience on GCP. We expect other providers will want to do the same and they are free to do so. Alibaba for example is customizing the Kubeflow deployment to using docker images mirror'd on their cloud to improve performance.

The ideal situation would be there would be non-vendor specific solution for GPUs in K8s but that's not the case.

**Comment by bhack** [comment]
Yes I understand you point of view as GCP member but so in that case who will care of bare metal kubernetes?  Bare metal is not a vendor I suppose.

>The ideal situation would be there would be non-vendor specific solution for GPUs in K8s but that's not the case.

I agree with you that it is ideal. But the kubernetes docs upstream status now is the only solution that we have and it is still making difference between GCP and others).

**Comment by vishh** [comment]
Driver installation is distro and environment (primarily legal) specific.
Only Nvidia can truly simplify that and they are working on it as mentioned
in
https://github.com/GoogleCloudPlatform/container-engine-accelerators/pull/51#issuecomment-399187623
I agree with Jeremy that it is difficult to abstract that in kubeflow.
Bringing up k8s is non-trivial. Kubeflow relies on k8s to be up and
running. Similarly, KF should rely on GPUs to be provisioned and managed as
part of k8s layer.

**Comment by bhack** [comment]
@vishh Yes I've already linked that comment in [this issue](https://github.com/kubeflow/kubeflow/issues/288#issuecomment-399222716)
I agree but actually, in k8s, we document ["officially"](https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/) two solution: GKE/GCE and "the others".

It is hard to integrate in this repository all the others k8s nodes GPU solutions:
* [kops gpu](https://github.com/kubernetes/kops/issues/518)
* [kubespray gpu reqeust for kubeflow](https://github.com/kubernetes-incubator/kubespray/issues/2079)
* [kube-aws](https://github.com/kubernetes-incubator/kube-aws/pull/1222), 
* [terraform](https://github.com/terraform-providers/terraform-provider-kubernetes/issues/149)
* [minikube](https://github.com/kubernetes/minikube/issues/2115)
* etc. etc. etc. 

So probably only GCP will have first class support in the repository if we merge the daemonset cause how we will integrate all these solutions untill nivida help us to find a solution to integrate it upstream in k8s?

**Comment by bhack** [comment]
Oh.. sorry for the noise you have already merge the GKE solution some hours ago. What documentation we will add i.e. for https://github.com/kubeflow/kubeflow/issues/32?

**Comment by jlewi** [comment]
@bhack I responded on #32 (this [comment](https://github.com/kubeflow/kubeflow/issues/32#issuecomment-403247872)).

You raise a good point about bare metal; feel free to open an issue about bare metal. I don't have much experience about deploying on bare metal so I'm not sure where to start in terms of getting a good bare metal deployment story.

Closing this issue because the deploy.sh script for GKE (see #1111) will now create the deployment set.
